plugins/dfu/dfu-device.c

```diff
@@ -1227,6 +1228,51 @@ dfu_device_open (FuUsbDevice *device, GError **error)
 		priv->status = DFU_STATUS_OK;
 	}
 
+	/* GD32VF103 encodes the serial number in UTF-8 (rather than UTF-16)
+	 * and also uses the first two bytes as the model identifier */
+	if (fu_device_has_custom_flag (FU_DEVICE (device), "gd32")) {
+#if G_USB_CHECK_VERSION(0,3,6)
+		const guint8 *buf;
+		gsize bufsz = 0;
+		guint16 langid = G_USB_DEVICE_LANGID_ENGLISH_UNITED_STATES;
+		guint8 idx = g_usb_device_get_serial_number_index (usb_device);
+		g_autofree gchar *chip_id = NULL;
+		g_autofree gchar *serial_str = NULL;
+		g_autoptr(GBytes) serial_blob = NULL;
+		serial_blob = g_usb_device_get_string_descriptor_bytes (usb_device,
+									idx,
+									langid,
+									error);
+		if (serial_blob == NULL)
+			return FALSE;
+		if (g_getenv ("FWUPD_DFU_VERBOSE") != NULL)
+			fu_common_dump_bytes (G_LOG_DOMAIN, "GD32 serial", serial_blob);
+		buf = g_bytes_get_data (serial_blob, &bufsz);
+		if (bufsz < 2) {
+			g_set_error_literal (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_NOT_SUPPORTED,
+					     "GD32 serial number invalid");
+			return FALSE;
+		}
+
+		/* ID is first two bytes */
+		chip_id = g_strdup_printf ("%02x%02x", buf[0], buf[1]);
+		dfu_device_set_chip_id (self, chip_id);
+
+		/* serial number follows */
+		serial_str = g_strndup ((const gchar *) buf + 2, bufsz - 2);
+		fu_device_set_serial (FU_DEVICE (device), serial_str);
+#else
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "GUsb version %s too old to support GD32",
```

Comment:
thinking like a user that would encounter this, perhaps you want to tell them the minimum version they need in the error?

Reply:
something like

			     "GUsb version %s too old to support GD32, "
			     "fwupd needs to be rebuit against 0.3.6",

Reply:
"...or later"

Reply:
ahh yes, good catch. Fixed.